File: src/main/java/org/dukcode/ps/codetree/trail02/chapter05/lesson02/IntroTheMomentWeMeet.java
package org.dukcode.ps.codetree.trail02.chapter05.lesson02;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.StringTokenizer;

/**
 * @see <a
 * href="https://www.codetree.ai/ko/trails/complete/curated-cards/intro-the-moment-we-meet/description">LINK</a>
 */
public class IntroTheMomentWeMeet {

  private static final int RANGE = 1_000_000;

  private static BufferedReader br;
  private static BufferedWriter bw;
  private static StringTokenizer st;

  public static void main(String[] args) throws IOException {
    br = new BufferedReader(new InputStreamReader(System.in));
    bw = new BufferedWriter(new OutputStreamWriter(System.out));

    st = new StringTokenizer(br.readLine());
    int n = Integer.parseInt(st.nextToken());
    int m = Integer.parseInt(st.nextToken());

    int[] posA = new int[2 * RANGE + 1];
    int[] posB = new int[2 * RANGE + 1];

    int time = Math.min(recordPos(posA, n), recordPos(posB, m));

    int meetTime = -1;
    for (int i = 1; i < time; i++) {
      if (posA[i + RANGE] == posB[i + RANGE]) {
        meetTime = i;
        break;
      }
    }

    bw.write(String.valueOf(meetTime));

    br.close();
    bw.close();

  }

  private static int recordPos(int[] pos, int n) throws IOException {
    int p = 0;
    int time = 1;
    for (int i = 0; i < n; i++) {
      st = new StringTokenizer(br.readLine());
      int dir = st.nextToken().equals("L") ? -1 : 1;
      int t = Integer.parseInt(st.nextToken());

      for (int now = time; now < time + t; ++now) {
        p += dir;
        pos[now + RANGE] = p;
      }

      time += t;
    }

    return time;
  }

}

File: src/main/kotlin/org/dukcode/ps/codetree/trail02/chapter05/lesson02/IntroTheMomentWeMeet.kt
package org.dukcode.ps.codetree.trail02.chapter05.lesson02.introTheMomentWeMeet

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.StringTokenizer

private val br = BufferedReader(InputStreamReader(System.`in`))
private val bw = BufferedWriter(OutputStreamWriter(System.out))

private const val MX = 1_000_000

/**
 * @see <a href="https://www.codetree.ai/ko/trails/complete/curated-cards/intro-the-moment-we-meet/description">LINK</a>
 */
fun main() {
    val (n, m) = parseInts()

    val a = buildPositions(n)
    val b = buildPositions(m)

    bw.write(findFirstMeetingSeconds(a, b).toString())

    bw.flush()
}

private fun buildPositions(n: Int): IntArray {
    var second = 1
    val positions = IntArray(MX + 1) { Int.MAX_VALUE }

    positions[0] = 0
    repeat(n) {
        val st = StringTokenizer(br.readLine())
        val d = if (st.nextToken() == "L") -1 else 1
        var t = st.nextToken().toInt()

        while (t-- > 0) {
            positions[second] = positions[second - 1] + d
            second++
        }
    }

    return positions
}

private fun findFirstMeetingSeconds(
    a: IntArray,
    b: IntArray,
): Int {
    for (second in 1..MX) {
        if (a[second] == Int.MAX_VALUE || b[second] == Int.MAX_VALUE) {
            return -1
        }

        if (a[second] == b[second]) {
            return second
        }
    }

    return -1
}

private fun parseInts() =
    StringTokenizer(br.readLine()).run {
        IntArray(countTokens()) {
            nextToken().toInt()
        }
    }
